src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs

```diff
@@ -94,28 +145,52 @@ protected internal XmlDocument(XmlImplementation imp) : base()
             _implementation = imp;
             _domNameTable = new DomNameTable(this);
 
-            // force the following string instances to be default in the nametable
-            XmlNameTable nt = this.NameTable;
-            nt.Add(string.Empty);
-            strDocumentName = nt.Add("#document");
-            strDocumentFragmentName = nt.Add("#document-fragment");
-            strCommentName = nt.Add("#comment");
-            strTextName = nt.Add("#text");
-            strCDataSectionName = nt.Add("#cdata-section");
-            strEntityName = nt.Add("#entity");
-            strID = nt.Add("id");
-            strNonSignificantWhitespaceName = nt.Add("#whitespace");
-            strSignificantWhitespaceName = nt.Add("#significant-whitespace");
-            strXmlns = nt.Add("xmlns");
-            strXml = nt.Add("xml");
-            strSpace = nt.Add("space");
-            strLang = nt.Add("lang");
-            strReservedXmlns = nt.Add(XmlReservedNs.NsXmlNs);
-            strReservedXml = nt.Add(XmlReservedNs.NsXml);
-            strEmpty = nt.Add(String.Empty);
+            strXmlns = Xmlns;
+            strXml = Xml;
+            strReservedXmlns = XmlReservedNs.NsXmlNs;
+            strReservedXml = XmlReservedNs.NsXml;
             baseURI = String.Empty;
-
             objLock = new object();
+
+            // If the implementation is using the framework NameTable avoid re-computing the hash codes. If not fallback
+            // we have to use the Add method exposed via XmlNameTable abstract class.
+            if (imp.NameTable is NameTable nt)
+            {
+                strDocumentName = nt.GetOrAddEntry(s_nameTableSeeds[DocumentNameSeedIndex].key, s_nameTableSeeds[DocumentNameSeedIndex].hash);
```

Comment:
I wonder if it wouldn't be cleaner to create a single instance of XmlNameTable and then just clone it (we might need to add copy constructor or clone method). No strong opinion either way though.